ubuntu/base/docker-install.sh

```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+set -ex; \
+
+apt-get update \
+&& apt-get install -q -y \
+    wget \
+    curl \
+    python \
+    lsb-release \
+    apt-transport-https \
+    ca-certificates \
+    software-properties-common \
+&& apt-get clean && rm -rf /var/lib/apt/lists/*
+
+#DOCKER_CE_VERSION="$(curl -s https://docs.docker.com/release-notes/docker-ce/ | grep 'h2' | grep "$DOCKER_VERSION" | cut -d '>' -f2 | cut -d '-' -f1)";
+#DOCKER_ENGINE_VERSION="$(curl -s https://docs.docker.com/release-notes/docker-engine/ | grep 'h2' | grep "$DOCKER_VERSION" | cut -d '>' -f2 | sed 's/[[:space:]].*//')";
+# Test version is no support('rc' version)
+
+if [[ "$DOCKER_VERSION" == *"ce"* ]]; then \
+    curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add -;
+    apt-key fingerprint 0EBFCD88; \
+    add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable edge"; \
+    DOCKER_LATEST_VERSION=$(curl -s https://api.github.com/repos/docker/docker-ce/releases/latest | grep "tag_name" | grep -o -E '[0-9].*' | cut -d '-' -f1); \
+    apt-get update; \
+    if [[ "$DOCKER_VERSION" == "latest" ]] || [[ "$DOCKER_VERSION" == "$DOCKER_LATEST_VERSION" ]]; then \
+        apt-get install -q -y docker-ce; \
+    else
+        DOCKER_CE_NAME=$(echo $DOCKER_VERSION | awk '{gsub(/-/,"~")}1'); \
+        INSTALL_VERSION=$(apt-cache madison docker-ce | grep "$DOCKER_CE_NAME" | head -1 | cut -d '|' -f2 | grep -o -E '[0-9].*'); \
+        apt-get install -q -y docker-ce=$INSTALL_VERSION; \
+    fi; \
+
+elif [[ "$DOCKER_VERSION" != *"ce"* ]]; then \
+    apt-key adv --keyserver hkp://ha.pool.sks-keyservers.net:80 --recv-keys 58118E89F3A912897C070ADBF76221572C52609D; \
+    mkdir -p /etc/apt/sources.list.d; \
+    echo "deb https://apt.dockerproject.org/repo ubuntu-`lsb_release -cs` main" | tee /etc/apt/sources.list.d/docker.list; \
+    apt-get update; \
+    DOCKER_INSTALL_ENGINE_VERSION="$(apt-cache madison docker-engine | grep "$DOCKER_VERSION" | head -1 | cut -d '|' -f2 | grep -o -E '[0-9].*')";
+    apt-get install -q -y docker-engine=$DOCKER_INSTALL_ENGINE_VERSION; \
+else
+    echo 'unknown version'
```

Comment:
I doubt that this else will never happen, is that correct?

Reply:
this is correct! this is to prevent if user forget to use '-ce' for docker-ce version. Because some version can be install without '-ce'!

Reply:
```
if [[ "$DOCKER_VERSION" == *"ce"* ]]; then \
    echo "ce version"

elif [[ "$DOCKER_VERSION" != *"ce"* ]]; then \
    echo "not ce version"
else
    echo 'what version?'
fi;
```

Reply:
@hieptranquoc give me an example of DOCKER_VERSION that the above code should `echo 'what version?'`

Reply:
@hoatle 17.03.x > 17.05.x! if user use `17.03.0` instead of `17.03.0-ce`, it can be installed. `https://apt.dockerproject.org/repo/pool/main/d/docker-engine/`

Reply:
fixed!